dependencies.yaml

```
@@ -850,19 +886,26 @@ dependencies:
     specific:
       - output_types: [requirements, pyproject]
         matrices:
-          - matrix: {cuda: "12.*"}
+          - matrix:
+              cuda: "12.*"
+              cuda_suffixed: "true"
             packages:
               - pylibcugraphops-cu12==24.8.*,>=0.0.0a0
-          - matrix: {cuda: "11.*"}
+          - matrix:
+              cuda: "11.*"
+              cuda_suffixed: "true"
             packages:
               - pylibcugraphops-cu11==24.8.*,>=0.0.0a0
-          - {matrix: null, packages: [*pylibcugraphops_conda]}
+          - {matrix: null, packages: [*pylibcugraphops_unsuffixed]}
 
   depends_on_cupy:
     common:
       - output_types: conda
         packages:
           - cupy>=12.0.0
+    # NOTE: this is not broken into groups by a 'cuda_suffixed' selector like
+    #       other packages with -cu{nn}x suffixes in this file because devcontainers-based
+    #       builds that build it from source expect a package with a `-cuda{nn}x` suffix
```

> that build it from source

The antecedent is a little unclear here. "it" refers to cugraph, not cupy here. We never build cupy from source in RAPIDS. Maybe we can simplify the comment like:

```suggestion
    # cupy is not built from source in devcontainers, so cuda_suffixed isn't needed
```

---

My choice to say "devcontainers-based builds" was a bad one. It's very unclear.

I should have said "DLFW builds".

We do not build `cupy` from source in RAPIDS devcontainers, but we DO build it from source in the RAPIDS DLFW builds. And there, pass the following through to `pip wheel`.

```text
--build-option="--cupy-package-name=cupy-cuda$(cut -d'.' -f1 <<< "${CUDA_VERSION}")x"
```

Here's where `cupy` exposes that option in its configuration:

https://github.com/cupy/cupy/blob/9381230cc55a5bf96bf838a6c6cc077b38fe68d1/install/cupy_builder/_context.py#L81-L83

So it's important, in the DFLW context, that `cugraph` depend on e.g. `"cupy-cuda12x"`, not `"cupy"`.

Maybe we should change that in DLFW in the future, or maybe the `cupy` dependency should be broken out into a separate block called like `depends_on_cupy` here in `dependencies.yaml`, so that these `cuda_suffixed: "false"` entries could be removed.

But with code freeze for 24.08 so close, I wasn't looking to change any behavior... just to reflect the current state in the smallest, least controversial changeset possible to get these merged in.

Given all that... would you support leaving this as-is and instead having a comment like the one I'm proposing on `cuml`'s `cupy` dependency over in https://github.com/rapidsai/cuml/pull/5974?

```text
# NOTE: cupy still has a "-cuda12x" suffix here, because it's suffixed
#       in DLFW builds
```

---

I think we are on the same page with everything but the wording of this comment -- the behavior here is what I expect, but I don't think the comment needs to describe _only_ DLFW.

To the best of my knowledge, there is no such pip package called `cupy` (public or internal/DLFW), so there's _never_ a case where we want cupy to be unsuffixed in `requirements` or `pyproject` lists. `cupy` is only ever used as the conda package name.

Note that for RAPIDS, devcontainers (including outside of DLFW) need the unsuffixed name [because that matches the exclusion list](https://github.com/rapidsai/devcontainers/blob/541895260fa9d3d772989c55c64a3ab98323f1c4/features/src/rapids-build-utils/opt/rapids-build-utils/bin/make-pip-dependencies.sh#L177-L186).

So either way, nothing here is DLFW-specific, if I understand correctly. We should be able to reword this to not mention DLFW and just say devcontainers. Perhaps the "from source" piece of my previous suggestion is irrelevant.

---

Think about it and merge whatever you think fits best for the comment. The behavior is what I expect, so I don't want to get stuck on the comment.

---

ok thanks!

I was thinking `cugraph` went into code freeze today, but I realize that wasn't correct. So we do have a bit more time with this one to make it look the way we long-term want it to. I'll push a change here (tomorrow) with a separated-out `depends_on_cupy` list and a less-confusing comment.

---

> need the unsuffixed name [because that matches the exclusion list](https://github.com/rapidsai/devcontainers/blob/541895260fa9d3d772989c55c64a3ab98323f1c4/features/src/rapids-build-utils/opt/rapids-build-utils/bin/make-pip-dependencies.sh#L177-L186).

And just to note... the exclusion list there includes both the suffixed and unsuffixed versions. That's what this line there does:

```bash
pip_noinstall+=("${pkg}" "${pkg}-cu.*");
```

---

To clarify, cupy is not captured by that exclusion list -- that list is constructed from RAPIDS packages (things in `manifest.yaml`). cupy is handled in a special case [here](https://github.com/rapidsai/devcontainers/blob/541895260fa9d3d772989c55c64a3ab98323f1c4/features/src/rapids-build-utils/opt/rapids-build-utils/bin/make-pip-dependencies.sh#L197), where we transform it into:
`cupy-cuda[0-9]+x` -> `cupy-cuda${cuda_version_major}x`

This is to work around issues in previous releases' `dependencies.yaml` that always output `cupy-cuda11x` instead of outputting the correct version based on the `cuda` matrix entry.

---

Thanks for clarifying that, appreciated!

---

I just pushed, updating the comment to this:

```yaml
# NOTE: This is intentionally not broken into groups by a 'cuda_suffixed' selector like
#       other packages with -cu{nn}x suffixes in this file.
#       All RAPIDS wheel builds (including in devcontainers) expect cupy to be suffixed.
```

Hopefully that more clearly captures the intent.

> *...maybe the `cupy`dependency should be broken out into a separate block called like depends_on_cupy here in `dependencies.yaml`*

🤦🏻 that's already how it works here in `cugraph`, sorry... too many very-similar-looking PRs yesterday.